cpp/src/arrow/util/io_util.cc

```diff
@@ -193,34 +210,79 @@ Status StringToNative(const std::string& s, NativePathString* out) {
   return Status::OK();
 }
 
+#if _WIN32
+Status NativeToString(const NativePathString& ws, std::string* out) {
+#if _WIN32
+  std::string s;
+  RETURN_NOT_OK(::arrow::util::WideStringToUTF8(ws, &s));
+  *out = std::move(s);
+#else
+  *out = ws;
+#endif
+  return Status::OK();
+}
+#endif
+
+#if _WIN32
+const wchar_t kNativeSep = L'\\';
+const wchar_t kGenericSep = L'/';
+const wchar_t* kAllSeps = L"\\/";
+#else
+const char kNativeSep = '/';
+const char kGenericSep = '/';
+const char* kAllSeps = "/";
+#endif
+
+NativePathString NativeSlashes(NativePathString s) {
+  return ReplaceChars(std::move(s), kGenericSep, kNativeSep);
+}
+
+NativePathString GenericSlashes(NativePathString s) {
+  return ReplaceChars(std::move(s), kNativeSep, kGenericSep);
+}
+
+NativePathString NativeParent(const NativePathString& s) {
+  // XXX What about trailing separators?
```

Comment:
Trailing separators should be ignored; `/a/b/` is equivalent to `/a/b` (and `/a//b//`). Also add tests for this

Reply:
Yes, will do.